src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/RazorHtmlCSharpLanguageServerClient.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.Utilities;
+using Nerdbank.Streams;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    [ClientName(ClientName)]
+    [Export(typeof(ILanguageClient))]
+    [ContentType(RazorLSPContentTypeDefinition.Name)]
+    internal class RazorHtmlCSharpLanguageServerClient : ILanguageClient
+    {
+        // ClientName enables us to turn on-off the ILanguageClient functionality for specific TextBuffers of content type RazorLSPContentTypeDefinition.Name.
+        // This typically is used in cloud scenarios where we want to utilize an ILanguageClient on the server but not the client; therefore we disable this
+        // ILanguageClient infrastructure on the guest to ensure that two language servers don't provide results.
+        public const string ClientName = "RazorLSPClientName";
+        private readonly IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> _requestHandlers;
+
+        [ImportingConstructor]
+        public RazorHtmlCSharpLanguageServerClient([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            if (requestHandlers is null)
+            {
+                throw new ArgumentNullException(nameof(requestHandlers));
+            }
+
+            _requestHandlers = requestHandlers;
+        }
+
+        public string Name => "Razor Html & CSharp Language Server Client";
+
+        public IEnumerable<string> ConfigurationSections => null;
+
+        public object InitializationOptions => null;
+
+        public IEnumerable<string> FilesToWatch => null;
+
+        public event AsyncEventHandler<EventArgs> StartAsync;
+        public event AsyncEventHandler<EventArgs> StopAsync
+        {
+            add { }
+            remove { }
+        }
+
+        public Task<Connection> ActivateAsync(CancellationToken token)
+        {
+            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+
+            _ = new RazorHtmlCSharpLanguageServer(serverStream, serverStream, _requestHandlers);
```

Comment:
Why pass the same values twice for `serverStream` and `clientStream` below this?

Reply:
`serverStream` represents one of the bidirectional streams to be used so a server can receive notifications and send notifications. Basically if the server is writing to the stream it's pushing data down to the client and if it's reading it's being pushed-to. 

Similarly `clientStream` represents a bidirectional stream and when the client writes to the `clientStream` behind the covers the "write" will actually be pushed to the server stream. 

To be frank i'm not entirely sure why it's factored this way but this is the approach that Roslyn took so we're following suite.

Reply:
fwiw the Roslyn version was based on the [C++ one](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fvc%2Fdesigntime%2FLanguageServer%2FLSP%2FLanguageClient.cs&version=GBmaster&line=51&lineEnd=52&lineStartColumn=1&lineEndColumn=1&lineStyle=plain).